### ISSUE TITLE
Add option for allowing overlapping AS numbers in different RIRs

### ISSUE THREAD

**Issue by indy-independence** [netbox-community/netbox#17133]
### NetBox version

v4.0.8

### Feature type

Data model extension

### Proposed functionality

Add a new flag on the RIR objects that allows disabling of the enforcement that AS numbers has to be globally unique

As requested in #15172 etc

### Use case

Private AS numbers are not globally unique and for example two different tenants might use the same private AS number for two different BGP islands, so I want Netbox to be able to have these two separate instances in different records in the database. In my use-case I basically need overlapping ASNs between different tenants, but allowing overlapping between different RIRs and creating one (or more) private RIRs per tenant is probably more flexible

### Database changes

New field on RIR class "enforce_unique_asns" which defaults to true and has same behavior as today, but unchecking should allow different RIRs to have overlapping ASNs

Add new field to ASN model called something like "rir_scope", could just be an integer field defaulting to 0 for the default behavior, or the integer could be refering to the id of the RIR if the enforcement of globally unique ASNs has been disabled? 
ASNField should not be unique on it's own, but combination of ASN + "rir_scope" should be unique

A problem with this is that you configure the uniqueness flag on the RIR, but then it's actually the ASN records that needs to be updated. So if you change the flag on an existing RIR, for consistency it would need to look up all existing ASNs in that RIR and change the "rir_scope" field etc. Are there better ways to solve this? (Instead of adding a new field to ASN, just remove the unique flag for ASN and then manually check uniqueness in code based on RIR setting before inserting/updating DB maybe?)

### External dependencies

None

**Comment by arthanson** [comment]
I think there would probably need to be discussion around this, what the actual implementation should be.

**Comment by indy-independence** [comment]
Yes it's good if we can agree how the implementation should work before starting to code, I kind of suggested two different ways of doing the unique check:
1. new "rir_scope" field on ASN model, and code for updating ASN records when RIR uniqueness flag is modified
2. code for doing uniqueness-check att all places adding/modifying AS-numbers
3. other?

**Comment by indy-independence** [comment]
It seems like ASN ranges also counts number of ASes, so we need to investigate if that check only considers things in it's own RIR or if it needs to be modified somehow to work with this overlapping AS feature

**Comment by jeremystretch** [comment]
This change needs to be tied to a new minor version as it would alter a uniqueness constraint in the database.

**Comment by dmulyalin** [comment]
If I may suggest, I believe ASN uniqueness should be tied to ASN Ranges and not RIRs as it is a more natural way of enforsing ASN uniqueness, similar to [how its done for vlans and vlan groups](https://github.com/netbox-community/netbox/blob/dab27695b9832dc64a26589d1787f34302a3f600/netbox/ipam/models/vlans.py#L205), and RIRs already can have multiple ranges associated with it.

Implementation details:

- link [ASN model](https://github.com/netbox-community/netbox/blob/dab27695b9832dc64a26589d1787f34302a3f600/netbox/ipam/models/asns.py#L94) with ASN Range using ``models.ForeignKey`` field
- update [ASN uniqueness constraint](https://github.com/netbox-community/netbox/blob/dab27695b9832dc64a26589d1787f34302a3f600/netbox/ipam/models/asns.py#L107) to take ASN Range into account using ``unique_together = ('asn_range', 'asn')`` constraint
- enhance RIR view "related objects" to have "ASN Ranges" reference

This approach is simpler to implement and provides greater flexibility as will allow to have duplicate ASN values recorded even within same RIR, for example if RIR is mapped to a company, and company manages customer networks that use private ASN, we can create two independent ranges for each customer and map them to single RIR.

Another use case, Company has multiple loosely coupled networks (corp networks, DataCentre networks, backbone network) managed by different departments with non overlapping routing domains, each of those network uses private ASN ranges, in that case to document such a scenario we can define multiple ranges for each network and link that with Company as a RIR.

Another use case, overtime multiple companies merged and their networks come together, it happens that private ASN ranges been used across both infrastructures, to capture that in Netbox we can again, define multiple ranges each pertaining to different private ASN usage and link the to single or multiple RIRs.

In addition, the flexibility of this approach would allow to document overlapping ASN for Tenants as well, for example we can have single RIR called ``PRIVATE_ASN`` linked with overlapping ranges and each range can be linked with different tenants. Where approach of using RIR to ease ASN uniqueness constraint would not allow to do that, since RIR model not linked with Tenants.

Moreover, using ASN Range as a uniqueness constraint for ASN opens the road for further enhancing solution by defining scope attribute for ASN range to use it in a similar way as how vlan groups use it.

@indy-independence @jeremystretch @arthanson what do you think?

**Comment by ragzilla** [comment]
Since there's some activity in a related issue which might come over here soon, I'd echo moving the uniqueness constraint into ASN ranges. ASN ranges can already associate to tenants to satisfy the data connectivity requested in #21137, but uniqueness on the ASN range allows more flexible usage. This should probably be gated behind a bool on the range (for legacy/new behavior), and then a hidden database field which can be left null for un-range'd ASNs, or ASNs in ranges without the bool set, and set to the range if the bool is set. This lets the uniqueness constraint in the DB apply to the combination of range_hidden,asn. This would eliminate the need for a database migration as all existing entries would end up in the default group.

An alternative, maybe cleaner implementation, could be to add ASN Groups, and just add ForeignKey linkages on both ASNs and ASNRanges to ASNGroups, then that would become the uniqueness constraint. Less need for a bool and hidden fields that way. But then you have two grouping objects for ASNs, not sure if that's desirable (or if this has been proposed and shot down before).

The use case I'm particularly interested in, is documented underlay ASNs for overlay fabrics. Within each fabric island we'll reuse RFC6996 ASNs in a deterministic way (so a particular spine or leaf in a repeatable fabric, will use the same ASN).

**Comment by pheus** [comment]
After some discussion, I think there are four plausible ways to approach this FR, and I’d really appreciate community feedback on which model feels most natural for NetBox before I take one of them further.

1. **RIR-scoped uniqueness**
   - Add something like `RIR.enforce_unique_asns`
   - Use RIR as the uniqueness boundary for overlapping ASNs
   - This is the closest match to the issue as filed, but it does mean using `RIR` as more than just an allocation authority

2. **Tenant-scoped overlap for private-use ASNs**
   - Keep global uniqueness for public ASNs
   - Allow overlap for private-use ASNs when they are scoped by `tenant`
   - This is narrower than the original FR, but probably the smallest/most standards-aligned solution

3. **Introduce an `ASNGroup` model**
   - Similar in spirit to `VLANGroup`, but for ASN uniqueness scope
   - This feels cleaner than overloading `RIR` or assuming `tenant` is always the right scope
   - The downside is that it is a bigger change, and unlike VLANs we already have `ASNRange`, so we’d need to be careful not to create overlapping concepts

4. **Postpone until a future namespace feature**
   - Hold off on this until there is a broader IPAM / routing namespace concept
   - Potentially the cleanest long-term model
   - But it also turns this into a larger design problem and delays the immediate feature

At the moment, I think the main question is:

> What object should actually own the uniqueness scope for overlapping ASNs in NetBox?

Some specific feedback that would be helpful:

- Should overlap be limited to private-use ASNs only, or should any ASN be allowed to overlap once scoped?
- Does `tenant` feel like the right built-in scope, or is that too narrow?
- Does `ASNGroup` sound like the right abstraction, or would that just be a narrower version of a future namespace feature?
- If `ASNGroup` is preferred, should it be a minimal scope object in v1, or should it also tie into ASN ranges / next-available allocation?

I’d especially appreciate feedback from anyone already using NetBox to model reused private ASNs across customers, labs, or isolated routing domains.

**Comment by oceans1908** [comment]
I think option 1 will meet our needs. We are using Netbox for years and are running into this issue, where we manage multiple networks in the same Netbox, and are running into the same ASN in the multiple networks. We can than use a RIR per network for the private ASN range and a RIR for Public ASN numbers. Don't think you should limit Netbox users in uniqueness of ASN numbers, so private and public ASN numbers should not be forces to be unique in Netbox. I think tenant is too narrow. Is ASNGroup simular to RIR.

**Comment by dmulyalin** [comment]
Why not to use ASNRange as a uniqueness constraint for BGP ASNs for reason highlighted in above comments @pheus ?

**Comment by suened** [comment]
We are having the same issue at this moment and I found this issue.

To answer the questions from @pheus 

**Should overlap be limited to private-use ASNs only, or should any ASN be allowed to overlap once scoped?**
I think any ASN should be allowed to overlap, not just private. For example a lab environments that mirror production using real public ASNs. And there are likely more edge cases that are hard to predict upfront. Drawing the line at private ASN means hitting a wall the moment one of those cases shows up in practice and then the feature needs to be revisited anyway. 

**Does tenant feel like the right built-in scope, or is that too narrow?**
Tenant feels like the right scope. NetBox already uses tenant as the isolation boundary for overlapping IP space in merger and multi-customer scenarios. Using the same pattern for ASNs feels consistent. 

**Does ASNGroup sound like the right abstraction, or would that just be a narrower version of a future namespace feature?**
Tenant already covers this. ASNGroup feels like an extra object to maintain for a problem that tenant solves more naturally. If a broader namespace feature is on the roadmap, ASNGroup risks becoming redundant before it matures.

**If ASNGroup is preferred, should it be a minimal scope object in v1, or should it also tie into ASN ranges / next-available allocation?**
As I think tenant feels like a better place this shouldn't be needed.

**Comment by ragzilla** [comment]
Lean toward 3 (`ASNGroup`), but 1 or 2 would work if that's community consensus. 3 feels like it fits with the other models already used for overlaps.

To the specific questions:
**Should overlap be limited to private-use ASNs only, or should any ASN be allowed to overlap once scoped?**
The use cases I've experienced are private use only, however I don't think that needs to be limited by the software. VRFs don't limit themselves to RFC1918 only addresses, and someone may have a requirement to use the full ASN range in a fabric.

**Does tenant feel like the right built-in scope, or is that too narrow?**
I feel tenant could be workable, but potentially too narrow. If a user is maintaining multiple fabric islands for a single tenant, this means that tenant now needs multiple sub-tenants for each fabric. 

**Does ASNGroup sound like the right abstraction, or would that just be a narrower version of a future namespace feature?**
It's narrower than a generic namespace as proposed in 4, but then so are the other existing groupings (VLANGroup/VRF), so it's more feature-parity and even once namespaces come into play people may still be inclined to use single namespace with multiple groups. Or even groups within namespaces (depending on how structured), e.g. I namespace out an internal entity which has its own fabrics, I'd still potentially need to manage overlapping ASN ranges for them within their namespace necessitating the ASNGroup.

**If ASNGroup is preferred, should it be a minimal scope object in v1, or should it also tie into ASN ranges / next-available allocation?**
I could use combined group + range + next-available (on the range which is a member of the group) when setting up sparse ASN blocks for fabrics- assigning ASNs from bit-aligned blocks. But a minimal scope object in v1 would work, since we can get similar behavior with ASNGroup per block, and ideally a next-available endpoint on the ASNGroup for automation.

**Comment by lvo12** [comment]
Option 3 would best fit my usecase. Within 1 tennant I have overlap in ASN's. I would not restrict it to public/private ASN ranges.
